Bottom: 2a81bf72dc5c39da3a8d3829f794d04d0741332b
Top:    7e93c56ca35ea9a24ff811af9573108b5687e393
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2016-02-25 19:46:35 +0000

Refresh of fixup-z-heights-of-exported

---

diff --git a/src/hid/common/object3d.c b/src/hid/common/object3d.c
index 19e7ff5..3b48da6 100644
--- a/src/hid/common/object3d.c
+++ b/src/hid/common/object3d.c
@@ -254,8 +254,6 @@ get_contour_edge_n_is_round (PLINE *contour, int n)
     }
 #endif
 
-  return false;
-
   while (n > 0) {
     edge = edge->next; /* The VNODE structure is circularly linked, so wrapping is OK */
     n--;
@@ -696,14 +694,14 @@ object3d_from_board_outline (void)
 
 #ifdef REVERSED_PCB_CONTOURS
   objects = object3d_from_contours (board_outline,
-                                    -HACK_BOARD_THICKNESS + HACK_MASK_THICKNESS, /* Bottom */
-                                    0                     -HACK_MASK_THICKNESS,  /* Top */
+                                    -HACK_BOARD_THICKNESS, /* Bottom */
+                                    0                    ,  /* Top */
                                     board_appearance,
                                     top_bot_appearance);
 #else
   objects = object3d_from_contours (board_outline,
-                                     HACK_BOARD_THICKNESS / 2 - HACK_MASK_THICKNESS, /* Bottom */
-                                    -HACK_BOARD_THICKNESS / 2 + HACK_MASK_THICKNESS, /* Top */
+                                     HACK_BOARD_THICKNESS / 2, /* Bottom */
+                                    -HACK_BOARD_THICKNESS / 2, /* Top */
                                     board_appearance,
                                     top_bot_appearance);
 #endif
@@ -904,7 +902,23 @@ pv_mask_callback (const BoxType * b, void *cl)
   struct mask_info *info = cl;
   POLYAREA *np, *res;
 
-  if (!(np = CirclePoly (pv->X, pv->Y, pv->Mask / 2, NULL)))
+  if (!(np = PinPoly (pv, pv->Mask)))
+    return 0;
+
+  poly_Boolean_free (info->poly, np, &res, PBO_SUB);
+  info->poly = res;
+
+  return 1;
+}
+
+static int
+pv_drill_callback (const BoxType * b, void *cl)
+{
+  PinType *pv = (PinType *)b;
+  struct mask_info *info = cl;
+  POLYAREA *np, *res;
+
+  if (!(np = CirclePoly (pv->X, pv->Y, (pv->DrillingHole + 1) / 2, NULL)))
     return 0;
 
   poly_Boolean_free (info->poly, np, &res, PBO_SUB);
@@ -947,14 +961,14 @@ object3d_from_soldermask_within_area (POLYAREA *area, int side)
 
 #ifdef REVERSED_PCB_CONTOURS
   objects = object3d_from_contours (info.poly,
-                                    (side == TOP_SIDE) ? -HACK_MASK_THICKNESS : -HACK_BOARD_THICKNESS,                       /* Bottom */
-                                    (side == TOP_SIDE) ? 0                    : -HACK_BOARD_THICKNESS + HACK_MASK_THICKNESS, /* Top */
+                                    (side == TOP_SIDE) ? 0                   - HACK_COPPER_THICKNESS : -HACK_BOARD_THICKNESS - HACK_COPPER_THICKNESS - HACK_MASK_THICKNESS, /* Bottom */
+                                    (side == TOP_SIDE) ? HACK_MASK_THICKNESS - HACK_COPPER_THICKNESS : -HACK_BOARD_THICKNESS - HACK_COPPER_THICKNESS,                       /* Top */
                                     mask_appearance,
                                     NULL);
 #else
   objects = object3d_from_contours (info.poly,
-                                    (side == TOP_SIDE) ? -HACK_BOARD_THICKNESS / 2 + HACK_MASK_THICKNESS : HACK_BOARD_THICKNESS / 2,                       /* Bottom */
-                                    (side == TOP_SIDE) ? -HACK_BOARD_THICKNESS / 2                       : HACK_BOARD_THICKNESS / 2 - HACK_MASK_THICKNESS, /* Top */
+                                    (side == TOP_SIDE) ? -HACK_BOARD_THICKNESS / 2 - HACK_COPPER_THICKNESS                       : HACK_BOARD_THICKNESS / 2 + HACK_COPPER_THICKNESS + HACK_MASK_THICKNESS, /* Bottom */
+                                    (side == TOP_SIDE) ? -HACK_BOARD_THICKNESS / 2 - HACK_COPPER_THICKNESS - HACK_MASK_THICKNESS : HACK_BOARD_THICKNESS / 2 + HACK_COPPER_THICKNESS, /* Top */
                                     mask_appearance,
                                     NULL);
 #endif
@@ -1319,7 +1333,7 @@ compute_depth (int group)
   int min_copper_group;
   int max_copper_group;
   int num_copper_groups;
-  int middle_copper_group;
+//  int middle_copper_group;
   int depth;
 
   top_group = GetLayerGroupNumberBySide (TOP_SIDE);
@@ -1328,12 +1342,12 @@ compute_depth (int group)
   min_copper_group = MIN (bottom_group, top_group);
   max_copper_group = MAX (bottom_group, top_group);
   num_copper_groups = max_copper_group - min_copper_group;// + 1;
-  middle_copper_group = min_copper_group + num_copper_groups / 2;
+//  middle_copper_group = min_copper_group + num_copper_groups / 2;
 
   if (group >= 0 && group < max_group) {
     if (group >= min_copper_group && group <= max_copper_group) {
       /* XXX: IS THIS INCORRECT FOR REVERSED GROUP ORDERINGS? */
-      depth = -(group - middle_copper_group) * HACK_BOARD_THICKNESS / num_copper_groups;
+      depth = -(group - min_copper_group) * (HACK_BOARD_THICKNESS + HACK_COPPER_THICKNESS) / num_copper_groups;
     } else {
       depth = 0;
     }
@@ -1474,7 +1488,7 @@ pv_copper_callback (const BoxType * b, void *cl)
   struct copper_info *info = cl;
   POLYAREA *np, *res;
 
-  if (!(np = CirclePoly (pv->X, pv->Y, pv->Thickness / 2, NULL)))
+  if (!(np = PinPoly (pv, PIN_SIZE (pv))))
     return 0;
 
   poly_Boolean_free (info->poly, np, &res, PBO_UNITE);
@@ -1519,7 +1533,11 @@ object3d_from_copper_layers_within_area (POLYAREA *area)
 
   for (group = min_phys_group; group <= max_phys_group; group++) {
 
-    Coord depth = compute_depth (group);
+#ifdef REVERSED_PCB_CONTOURS
+    Coord depth = compute_depth (group) - HACK_BOARD_THICKNESS;
+#else
+    Coord depth = compute_depth (group) + HACK_BOARD_THICKNESS / 2;
+#endif
     info.poly = NULL;
 
     fprintf (stderr, "Computing copper geometry for group %i\n", group);
@@ -1537,7 +1555,7 @@ object3d_from_copper_layers_within_area (POLYAREA *area)
     END_LOOP;
 #endif
 
-#if 0
+#if 1
     fprintf (stderr, "Accumulating pin + via pads\n");
     r_search (PCB->Data->pin_tree, &bounds, NULL, pv_copper_callback, &info);
     r_search (PCB->Data->via_tree, &bounds, NULL, pv_copper_callback, &info);
@@ -1578,6 +1596,12 @@ object3d_from_copper_layers_within_area (POLYAREA *area)
    *
    */
 
+#if 1
+    fprintf (stderr, "Subtracting pin + via drills\n");
+    r_search (PCB->Data->pin_tree, &bounds, NULL, pv_drill_callback, &info);
+    r_search (PCB->Data->via_tree, &bounds, NULL, pv_drill_callback, &info);
+#endif
+
   if (info.poly == NULL)
     {
       fprintf (stderr, "Skipping layer group %i, info.poly was NULL\n", group);
@@ -1590,14 +1614,14 @@ object3d_from_copper_layers_within_area (POLYAREA *area)
                               depth,                         /* Bottom */
                               depth + HACK_COPPER_THICKNESS, /* Top */
 #else
-                              -depth + HACK_COPPER_THICKNESS, /* Bottom */
-                              -depth,                         /* Top */
+                              -depth,                         /* Bottom */
+                              -depth - HACK_COPPER_THICKNESS, /* Top */
 #endif
                               copper_appearance,
                               NULL));
 
 
-    break;
+//    break;
   }
 
 
diff --git a/src/hid/step/step.c b/src/hid/step/step.c
index a4fe3c5..2bb9bfd 100644
--- a/src/hid/step/step.c
+++ b/src/hid/step/step.c
@@ -152,7 +152,7 @@ step_do_export (HID_Attr_Val * options)
   if (filename == NULL)
     filename = "pcb-out.step";
 
-#if 0
+#if 1
   board_outline_list = object3d_from_board_outline ();
 #else
   board_outline_list = NULL;
@@ -202,11 +202,13 @@ step_do_export (HID_Attr_Val * options)
         poly_DelContour (&curc);
       }
 
-//    mask_objects = object3d_from_soldermask_within_area (piece, TOP_SIDE);
-//    board_outline_list = g_list_concat (board_outline_list, mask_objects);
+#if 1
+    mask_objects = object3d_from_soldermask_within_area (piece, TOP_SIDE);
+    board_outline_list = g_list_concat (board_outline_list, mask_objects);
 
-//    mask_objects = object3d_from_soldermask_within_area (piece, BOTTOM_SIDE);
-//    board_outline_list = g_list_concat (board_outline_list, mask_objects);
+    mask_objects = object3d_from_soldermask_within_area (piece, BOTTOM_SIDE);
+    board_outline_list = g_list_concat (board_outline_list, mask_objects);
+#endif
 
     copper_layer_objects = object3d_from_copper_layers_within_area (piece);
     board_outline_list = g_list_concat (board_outline_list, copper_layer_objects);
diff --git a/src/polygon.c b/src/polygon.c
index 54c3425..759fe75 100644
--- a/src/polygon.c
+++ b/src/polygon.c
@@ -1055,7 +1055,28 @@ Subtract (POLYAREA * np1, PolygonType * p, bool fnp)
 
 /* create a polygon of the pin clearance */
 POLYAREA *
-PinPoly (PinType * pin, Coord thick, Coord clear)
+PinPoly (PinType * pin, Coord thick)
+{
+  int size = (thick + 1) / 2;
+
+  if (TEST_FLAG (SQUAREFLAG, pin))
+    {
+      return RectPoly (pin->X - size, pin->X + size,
+                       pin->Y - size, pin->Y + size);
+    }
+  else if (TEST_FLAG (OCTAGONFLAG, pin))
+    {
+      return OctagonPoly (pin->X, pin->Y, thick);
+    }
+  else
+    {
+      return CirclePoly (pin->X, pin->Y, size, NULL);
+    }
+}
+
+/* create a polygon of the pin clearance */
+static POLYAREA *
+PinClearPoly (PinType * pin, Coord thick, Coord clear)
 {
   int size;
 
@@ -1119,7 +1140,7 @@ SubtractPin (DataType * d, PinType * pin, LayerType * l, PolygonType * p)
     }
   else
     {
-      np = PinPoly (pin, PIN_SIZE (pin), pin->Clearance);
+      np = PinClearPoly (pin, PIN_SIZE (pin), pin->Clearance);
       if (!np)
         return -1;
     }
@@ -1224,7 +1245,7 @@ pin_sub_callback (const BoxType * b, void *cl)
     }
   else
     {
-      np = PinPoly (pin, PIN_SIZE (pin), pin->Clearance);
+      np = PinClearPoly (pin, PIN_SIZE (pin), pin->Clearance);
       if (!np)
         longjmp (info->env, 1);
     }
diff --git a/src/polygon.h b/src/polygon.h
index 1343d03..f077f02 100644
--- a/src/polygon.h
+++ b/src/polygon.h
@@ -90,7 +90,7 @@ POLYAREA * CirclePoly (Coord x, Coord y, Coord radius, char *name);
 POLYAREA * OctagonPoly(Coord x, Coord y, Coord radius);
 POLYAREA * LinePoly(LineType *l, Coord thick, char *name);
 POLYAREA * ArcPoly(ArcType *l, Coord thick, char *name);
-POLYAREA * PinPoly(PinType *l, Coord thick, Coord clear);
+POLYAREA * PinPoly(PinType *l, Coord thick);
 POLYAREA * BoxPolyBloated (BoxType *box, Coord radius);
 POLYAREA * PadPoly (PadType *pad, Coord size);
 void frac_circle (PLINE *, Coord, Coord, Vector, int);
